### ISSUE TITLE
MediaItem Resolver

### ISSUE THREAD

**Issue by AlenToma** [martijn00/XamarinMediaManager#510]
## 🚀 Feature Requests

This is really importend request for my app.
I have a youtube manager app, I would like to have an `url` resolver or `MediaItem` resolver.
### Describe the feature
Ok say that i have a list of `MediaItem`,
with diffrrent URLs 
`https://www.youtube.com/watch?v=ZulwKDO-6B4&list=FLNp78z85NAdOtLgH16l-T7Q&index=27`

`XamarinMediaManager` should point to a `MediaItem` resolver, before starting to play the mediaItem. 
The resolver could then change URI/ decrypt it so the library could then play it.

This Feature should be a global one in the library.
I dont know of this kind of Feature is available right now as i have checked. 
### Why this is importend
Think if you have a list of videos, 20 or more. 
decrypting all those url  at once take to much time so i would like to do it the moment `XamarinMediaManager` try to play the `MediaItem`. 

If the `MediaItem` Url need decrypting then i would do so otherwise i would just return the same `MediaItem`

### Platforms affected (mark all that apply)
- [ ] :robot: Android
- [ ] :monkey: Xamarin.Forms

**Comment by martijn00** [comment]
It's not totally clear what you want to me. Do you want to intercept the url and change it into something else, or do you want to extract the metadata just before playing?

**Comment by AlenToma** [comment]
Yes, something like intercept the url, before the player start playing it.
As you can see above the library will not understand the youtube url. so i need to intercept and decrypt the url before it become playable.

**Comment by martijn00** [comment]
Ok, but why not put in the already decrypted url in into the player in the first place? And can't use use the https://github.com/martijn00/XamarinMediaManager/blob/develop/MediaManager/Player/IMediaPlayer.cs#L61 BeforePlaying event?

**Comment by AlenToma** [comment]
Thats it thx.
Did not know it exist. Now i could simple make change to `e.MediaItem.MediaUri`, and it will be applied right?

**Comment by martijn00** [comment]
Yes, but i'm not 100% sure this would be called early enough on Android. The implementation there is a bit different. Let me know if it works.

**Comment by AlenToma** [comment]
Ok will test it and let you know, before closing this request. thx agen.

**Comment by AlenToma** [comment]
I have checket `BeforePlaying` and its working but not as i mentioned.

`BeforePlaying` trigger only when the url is valid and playable. 

A Resolver work differently. it will check that an url is valid and also apply changes to it if necessary.

Will have to implement it on my own and let you have a look, and see if its a good thing to implement in the library.

**Comment by AlenToma** [comment]
I have been trying to implement a `Resolver ` but i cant find where the player start to play the item.

`BeforePlaying`, trigger after the player source has been set so making changes `e.MediaItem.MediaUri` dose not matter in this case.

Could you help ?

**Comment by DominicMaas** [comment]
I think what @AlenToma is describing is something similar to the UWP [MediaBinder](https://docs.microsoft.com/en-us/uwp/api/windows.media.core.mediabinder), (see the *Defer binding of media content for items in a playback list by using MediaBinder* heading under [this](https://docs.microsoft.com/en-us/windows/uwp/audio-video-camera/media-playback-with-mediasource) page - near the bottom).

Basically the idea is to add a list of x items to the playlist, depending on the audio source, it might not make sense to apply the URL straight away (for example with YouTube, the URL must be decoded which takes some time, so decoding 20 videos when starting a playlist is not a great experience).

Instead, getting the URL is deferred until later (done by invoking an event the user sets, maybe one of two songs ahead)?

Did you get anything working in the end @AlenToma?
